packages/server/modules/workspaces/tests/integration/workspaceJoinRequests.spec.ts

```diff
@@ -70,5 +85,89 @@ const { FF_WORKSPACES_MODULE_ENABLED } = getFeatureFlags()
         ).to.deep.equal({ status: 'dismissed' })
       })
     })
+
+    describe('requestToJoinWorkspaceFactory, returns a function that ', () => {
+      it('throws a NotFoundError if the user does not exists', async () => {
+        const err = await expectToThrow(() =>
+          requestToJoinWorkspaceFactory({
+            createWorkspaceJoinRequest: (async () =>
+              Promise.resolve()) as unknown as CreateWorkspaceJoinRequest,
+            sendWorkspaceJoinRequestReceivedEmail: async () => Promise.resolve(),
+            getUserById: async () => null,
+            getWorkspace: async () => null
+          })({ workspaceId: createRandomString(), userId: createRandomString() })
+        )
+
+        expect(err.message).to.equal('User not found')
+      })
+      it('throws a WorkspaceNotFoundError if the workspace does not exists', async () => {
+        const user = await createTestUser({})
+        const err = await expectToThrow(() =>
+          requestToJoinWorkspaceFactory({
+            createWorkspaceJoinRequest: (async () =>
+              Promise.resolve()) as unknown as CreateWorkspaceJoinRequest,
+            sendWorkspaceJoinRequestReceivedEmail: async () => Promise.resolve(),
+            getUserById: async () => user as unknown as UserWithOptionalRole,
+            getWorkspace: async () => null
+          })({ workspaceId: createRandomString(), userId: createRandomString() })
+        )
+
+        expect(err.message).to.equal(WorkspaceNotFoundError.defaultMessage)
+      })
+      it('creates a join request and sends an email to all admins', async () => {
+        const createWorkspaceJoinRequest = createWorkspaceJoinRequestFactory({ db })
+
+        const sendWorkspaceJoinRequestReceivedEmailCalls: Parameters<SendWorkspaceJoinRequestReceivedEmail>[number][] =
+          []
+        const sendWorkspaceJoinRequestReceivedEmail = async (
+          args: Parameters<SendWorkspaceJoinRequestReceivedEmail>[number]
+        ) => sendWorkspaceJoinRequestReceivedEmailCalls.push(args)
+
+        const user: BasicTestUser = {
+          id: '',
+          name: 'John Speckle',
+          email: createRandomEmail(),
+          role: Roles.Server.Admin,
+          verified: true
+        }
+
+        await createTestUser(user)
```

Comment:
why not reuse users/workspaces across tests? i get that you don't want cross-test pollution, but you can always account for that by not doing mutations (e.g. just reading the user, never changing it) or doing cleanup in after/afterEach